### DISCUSSION TITLE
Value of net not updating in Cocotb, but waveform says it does

### DISCUSSION THREAD

**Discussion by Paecificjr** [cocotb/cocotb#3397] [Q&A] [answered]
I have a memory module that I am driving with a force and that seems to work. I write the random data to the memory and it latches. The first assert passes. However, when I get to the second assert the value is incorrect. I figured there was something wrong, 

```python
for i in range(10):
    data = sample(range(0, 0xFFFFFFFF), 65536)
    dut.mem.value = data
    await RisingEdge(dut.clk)

    assert dut.mem.value == data
        
    count = 0
    while dut.o_data0.value != data[0] and count < 100:
        await ClockCycles(dut.clk, 1)

    assert dut.o_data0.value == data[0]
```

However, when I checked the waves, it appears that the value was updated. This signal continues like this forever. I left the `0xDEADBEEF` in there so that the transition could be seen. 
![cocotb_bug0](https://github.com/cocotb/cocotb/assets/8717175/f66d3170-92c9-4af7-95bc-950ca849786c)

I have other tests that drive the memory from the ports and then read it back. Those work fine. Is there something I am missing?

**Comment by Paecificjr** [comment]
So I changed the code to this and it worked. Not sure why though. Now the net seems to be reporting correctly. I would love some enlightenment though.

```python
for i in range(10):
    data = sample(range(0, 0xFFFFFFFF), 65536)
    dut.mem.value = data
    await RisingEdge(dut.clk)

    assert dut.mem.value == data
    
    await ClockCycles(dut.clk, 1)

    assert dut.o_data0.value == dut.mem[0].value
```

**Reply by ktbarrett** [reply, marked as answer]
The difference is you are waiting a clock cycle before testing `dut.o_data0.value` in this code, while in the original code you are testing `dut.o_data0.value` before your first `await ClockCycles(...)`. Just add `await ClockCycles(...)` before your while loop in the first code, you are miscounting your timing.

**Reply by Paecificjr** [reply]
That was it! Thanks. Is there a discord/irc that people use to ask general usage questions?

**Reply by ktbarrett** [reply]
There is a [gitter/matrix room](https://app.gitter.im/#/room/#cocotb_Lobby:gitter.im).
